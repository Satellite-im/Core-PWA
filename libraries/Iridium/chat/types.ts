export type ConversationMessage = {
  conversationId: string
  from: string
  body: string
  at: number
  type: 'text' | 'file' | 'group' | 'glyph' | 'divider' | 'string'
  attachments: string[]
}

export type Conversation = {
  id: string
  type: 'direct' | 'group'
  name?: string
  participants: string[]
  createdAt: number
  updatedAt: number
  messages: {
    [key: string]: ConversationMessage
  }
  messageArray: ConversationMessage[]
}

export const ChatError = {
  CONVERSATION_EXISTS: 'error.chat.conversation_exists',
  CONVERSATION_NOT_FOUND: 'error.chat.conversation_not_found',
  MESSAGE_NOT_SENT: 'error.chat.message_not_sent',
}
