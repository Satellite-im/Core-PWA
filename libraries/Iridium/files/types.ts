/* eslint-disable no-use-before-define */
import { DIRECTORY_TYPE } from '~/libraries/Files/types/directory'
import { FILE_TYPE } from '~/libraries/Files/types/file'

export type FileUpdate = 'upload' | 'delete' | 'like' | 'share'

interface Shared {
  id: string
  name: string
  liked: boolean
  shared: boolean
  modified: number
  size: number
  parentId: IridiumDirectory['id'] // empty string if root item
}

export interface IridiumFile extends Shared {
  type: FILE_TYPE
  description?: string
  thumbnail: string
  extension: string
  nsfw: boolean
}

export interface IridiumDirectory extends Shared {
  type: DIRECTORY_TYPE
  children: IridiumItem[]
}

export type IridiumItem = IridiumFile | IridiumDirectory
