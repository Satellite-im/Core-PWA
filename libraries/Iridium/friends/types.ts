import { User } from '../users/types'

export type Friend = User & {}

export type FriendRequestStatus =
  | 'pending'
  | 'accepted'
  | 'rejected'
  | 'removed'

export type FriendRequest = {
  did: User['did']
  incoming: boolean
  status: FriendRequestStatus
  at: number
}

export const FriendsError = {
  NETWORK_ERROR: 'errors.friends.network',
  FRIEND_EXISTS: 'errors.friends.exists',
  FRIEND_NOT_FOUND: 'errors.friends.not_found',
  REQUEST_NOT_FOUND: 'errors.friends.request_not_found',
  REQUEST_NOT_SENT: 'errors.friends.request_not_sent',
  REQUEST_ALREADY_SENT: 'errors.friends.request_already_sent',
}
