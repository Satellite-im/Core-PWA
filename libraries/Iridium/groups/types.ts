import { IridiumDocument } from '@satellite-im/iridium'
import Group from './Group'

export type GroupMemberDetails = {
  id: string
  name: string
  photoHash: string
}

export type GroupConfig = {
  name: string
  origin: string
  avatar?: string
  members: { [did: string]: GroupMemberDetails }
  metadata?: IridiumDocument
}

export type GroupData = GroupConfig & {
  id: string
}

export type GroupMap = { [key: string]: Group }

export type GroupManagerEvent =
  | 'group-created'
  | 'group-updated'
  | 'group-deleted'
  | 'group-member-added'
  | 'group-member-removed'
  | 'group-member-updated'
  | 'group-member-joined'
  | 'group-member-left'

export enum GroupsError {
  NETWORK_ERROR = 'errors.groups.network',
  NOT_A_MEMBER = 'errors.groups.not_a_member',
  GROUP_ALREADY_EXISTS = 'error.groups.group_already_exists',
  GROUP_NOT_FOUND = 'error.groups.group_not_found',
  GROUP_NOT_INITIALIZED = 'errors.groups.group_not_initialized',
  INVITE_ALREADY_SENT = 'errors.groups.request_already_sent',
  RECIPIENT_NOT_FOUND = 'errors.groups.recipient_not_found',
  USER_NOT_INITIALIZED = 'errors.groups.user_not_initialized',
}
