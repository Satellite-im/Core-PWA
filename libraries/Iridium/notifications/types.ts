import { IridiumPeerIdentifier } from '~/../iridium/dist'

export const NotificationsError = {
  NOTIFICATION_NOT_SENT: 'error.notifications.notifications_not_sent',
}

export enum NotificationType {
  FRIEND_REQUEST = 'friend_request',
  DIRECT_MESSAGE = 'direct_message',
  GROUP_MESSAGE = 'group_message',
  MENTION = 'mention',
  MEMBER_JOIN = 'member_join',
  MEMBER_LEAVE = 'member_leave',
  ADDED_TO_GROUP = 'added_to_group',
  GROUP_CONVERSATION_CREATED = 'group_conversation_created',
  CALL_INCOMING = 'call_incoming',
}

export type NotificationTypeValues = NotificationType[keyof NotificationType]

export type Notification<P = {}> = {
  at: number
  type: NotificationType
  senderId: IridiumPeerIdentifier
  title: string
  description: string
  seen: boolean
  payload: P
  id?: string
  image?: string
}

export type FriendRequestNotificationPayload = {
  senderId: IridiumPeerIdentifier
}

export type MessageNotificationPayload = {
  conversationId: string
  messageId: string
}

export type GroupConversationCreatedNotificationPayload = {
  conversationId: string
}

export type MemberJoinNotificationPayload = MessageNotificationPayload & {
  addedMemberIds: IridiumPeerIdentifier[]
}

export enum ActionTypeIds {
  CALL_INCOMING = 'call_incoming',
  DEFAULT = 'default',
  FRIEND_REQUEST = 'friend_request',
  MESSAGE = 'message',
}

export type NotificationPayloads =
  | FriendRequestNotificationPayload
  | GroupConversationCreatedNotificationPayload
  | MemberJoinNotificationPayload
  | MessageNotificationPayload

export type NotificationBase<P = {}> = Omit<
  Notification<P>,
  'title' | 'description'
>

export type NotifCreateHandler<P = {}> = {
  handler: (notif: P) => void
}

export enum NotifActionTypes {
  TAP = 'tap',
  RESPOND = 'respond',
  ACCEPT = 'accept',
  DECLINE = 'decline',
}

export type NotifActionHandler<P = {}> = {
  [key in NotifActionTypes]?: (payload: P, inputValue?: string) => void
}
