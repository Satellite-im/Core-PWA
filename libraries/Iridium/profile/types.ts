import { User } from '../friends/types'

export type Profile = User & {
  friends: Friend[]
  chats: Chat[]
  groups: Group[]
  notifications: Notification[]
  settings: IridiumDocument
}
