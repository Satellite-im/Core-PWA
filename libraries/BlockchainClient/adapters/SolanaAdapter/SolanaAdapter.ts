import {
  Keypair,
  PublicKey,
  RpcResponseAndContext,
  SignatureResult,
} from '@solana/web3.js'
import nacl from 'tweetnacl'
import nacl from 'tweetnacl'
import {
  Account,
  Adapter,
  CreateUserParams,
  FriendsEvents,
  User,
  Group,
} from '../../interfaces'
import {
  accountFromKeyapair,
  accountFromWallet,
  walletFromAccount,
} from './utils'
import SolanaManager from '~/libraries/Solana/SolanaManager/SolanaManager'
import { AccountsError } from '~/store/accounts/types'

export default class SolanaAdapter implements Adapter {
  private readonly solanaManager: SolanaManager

  constructor() {
    this.solanaManager = new SolanaManager()
    this.usersProgram = null
  }

  async signMessage(message: string): Promise<Uint8Array> {
    const messageBytes = Buffer.from(message, 'utf8')
    const wallet = await this.getPayerAccount()
    if (!wallet?.secretKey) {
      throw new Error(AccountsError.PAYER_NOT_PRESENT)
    }

    return nacl.sign.detached(messageBytes, wallet?.secretKey)
  }

  _getConnectionStatus(): boolean {
    if (this.solanaManager.isInitialized()) {
      return true
    }
    return false
  }

  get friendsProgram(): FriendsProgram {
    if (this._friendsProgram) return this._friendsProgram
    if (this.solanaManager.isInitialized()) {
      this._friendsProgram = new FriendsProgram(this.solanaManager)
      return this._friendsProgram
    }

    throw new Error(
      'Unable to get FriendsProgram instance: Solana not initialized yet',
    )
  }

  async createRandomAccount(): Promise<Account> {
    const wallet = await this.solanaManager.createRandomKeypair()
    return accountFromWallet(wallet)
  }

  async createUser(params: CreateUserParams): Promise<boolean> {
    if (params.account) {
      await this.initSolanaManager(params.account)
      await this.initUserProgram()
      await this.usersProgram?.create(
        params.name,
        params.photoHash,
        params.status,
      )
      return true
    }
    return false
  }

  async getAccountBalance(account: Account): Promise<number | null> {
    await this.initSolanaManager(account)
    return this.solanaManager.getCurrentAccountBalance()
  }

  async getAccountFromMnemonic(mnemonic?: string): Promise<Account | null> {
    if (!mnemonic) {
      throw new Error('Mnemonic is required when working with Solana directly')
    }
    await this.solanaManager.initializeFromMnemonic(mnemonic)
    const wallet = this.solanaManager.getMainSolanaWalletInstance()
    if (wallet) {
      return accountFromWallet(wallet)
    }
    return null
  }

  async requestAirdrop(): Promise<RpcResponseAndContext<SignatureResult> | null> {
    await this.solanaManager.requestAirdrop().then((out) => {
      return out
    })
    return null
  }

  private async initSolanaManager(account: Account) {
    await this.solanaManager.initializeFromSolanaWallet(
      walletFromAccount(account),
    )
  }

  private getAccounts(): {
    payerAccount: Keypair
    userAccount: Keypair
  } {
    const payerAccount = this.solanaManager.getActiveAccount()
    if (!payerAccount) {
      throw new Error(AccountsError.PAYER_NOT_PRESENT)
    }

    const userAccount = this.solanaManager.getUserAccount()
    if (!userAccount) {
      throw new Error(AccountsError.USER_DERIVATION_FAILED)
    }
    return {
      payerAccount,
      userAccount,
    }
  }

  async getUserInfo(address: string): Promise<User | null> {
    try {
      this.usersProgram = new UsersProgram(this.solanaManager)
      const e = await this.usersProgram.getUserInfo(address)
      return e
    } catch (e) {
      return null
    }
  }

  async getUsersInfo(addresses: string[]): Promise<User[]> {
    try {
      this.usersProgram = new UsersProgram(this.solanaManager)
      const users = await this.usersProgram.getUsersInfo(addresses)
      return users.map(
        (user) =>
          <User>{
            ...user,
          },
      )
    } catch (e) {
      throw new Error("Couldn't get users info")
    }
  }

  async getActiveAccount(): Promise<Account | undefined> {
    const account = this.solanaManager.getActiveAccount()
    return !account ? undefined : accountFromKeyapair(account)
  }

  async getPayerAccount(): Promise<Account | undefined> {
    const account = this.solanaManager.payerAccount
    return !account ? undefined : accountFromKeyapair(account)
  }
}
