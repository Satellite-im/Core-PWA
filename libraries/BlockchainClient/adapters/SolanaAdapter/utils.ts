import { Keypair, PublicKey } from '@solana/web3.js'
import { encode as base58encode, decode as base58decode } from 'micro-base58'
import { SolanaWallet } from '~/types/solana/solana'
import {
  Account,
  FriendStatus,
  Status,
} from '~/libraries/BlockchainClient/interfaces'

export const accountFromWallet = (wallet: SolanaWallet): Account => {
  return {
    privateKey: base58encode(wallet.keypair.secretKey),
    address: wallet.address,
    mnemonic: wallet.mnemonic,
    path: wallet.path,
  }
}

export const publicKeyFromAddress = (address: string): PublicKey => {
  return new PublicKey(base58decode(address))
}

export const walletFromAccount = (account: Account): SolanaWallet => {
  const keyPair = Keypair.fromSecretKey(base58decode(account.privateKey), {
    skipValidation: true,
  })
  return {
    keypair: keyPair,
    address: account.address,
    mnemonic: account.mnemonic,
    path: account.path,
  }
}
