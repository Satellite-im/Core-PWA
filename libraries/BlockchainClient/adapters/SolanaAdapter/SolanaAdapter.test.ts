import SolanaAdapter from '~/libraries/BlockchainClient/adapters/SolanaAdapter/SolanaAdapter'
import { AccountsError } from '~/store/accounts/types'

describe('Test Solana Adapter', () => {
  it('should initialize the constructor', () => {
    const konstruktor = new SolanaAdapter()
    expect(() => konstruktor).not.toThrow(Error)
  })
  it('should sign message but returns error', async () => {
    // because getPayerAccount is not mocked
    const konstruktor = new SolanaAdapter()

    try {
      const result = await konstruktor.signMessage('example message')
    } catch (error) {
      expect(error).toBeInstanceOf(Error)
      expect(error).toHaveProperty('message', AccountsError.PAYER_NOT_PRESENT)
    }
  })
  it('should get connection status before solanaManager is is initialized', () => {
    const konstruktor = new SolanaAdapter()

    const result = konstruktor._getConnectionStatus()
    expect(result).toBeFalsy()
  })
  it('should get(ter) friendsProgram and return error', () => {
    // because Solana is not initialized
    const konstruktor = new SolanaAdapter()

    try {
      const result = konstruktor.friendsProgram
    } catch (error) {
      expect(error).toBeInstanceOf(Error)
      expect(error).toHaveProperty(
        'message',
        'Unable to get FriendsProgram instance: Solana not initialized yet',
      )
    }
  })
})
