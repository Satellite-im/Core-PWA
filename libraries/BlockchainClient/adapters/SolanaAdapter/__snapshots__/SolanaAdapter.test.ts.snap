// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Solana Adapter should initialize the constructor 1`] = `
SolanaAdapter {
  "_friendsProgram": null,
  "groupsProgram": null,
  "solanaManager": SolanaManager {
    "accounts": Array [],
    "clusterApiUrl": "https://api.devnet.solana.com",
    "connection": Connection {
      "_blockhashInfo": Object {
        "lastFetch": 0,
        "latestBlockhash": null,
        "simulatedSignatures": Array [],
        "transactionSignatures": Array [],
      },
      "_commitment": "confirmed",
      "_confirmTransactionInitialTimeout": undefined,
      "_disableBlockhashCaching": false,
      "_nextClientSubscriptionId": 0,
      "_pollingBlockhash": false,
      "_rpcBatchRequest": [Function],
      "_rpcClient": ClientBrowser {
        "callServer": [Function],
        "options": Object {
          "generator": [Function],
          "notificationIdNull": false,
          "replacer": null,
          "reviver": null,
          "version": 2,
        },
      },
      "_rpcEndpoint": "https://api.devnet.solana.com",
      "_rpcRequest": [Function],
      "_rpcWebSocket": Client {
        "_events": Object {
          "accountNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "close": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "error": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "logsNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "open": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "programNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "rootNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "signatureNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "slotNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
          "slotsUpdatesNotification": EE {
            "context": [Circular],
            "fn": [Function],
            "once": false,
          },
        },
        "_eventsCount": 10,
        "address": "wss://api.devnet.solana.com/",
        "autoconnect": false,
        "current_reconnects": 0,
        "generate_request_id": [Function],
        "max_reconnects": Infinity,
        "queue": Object {},
        "ready": false,
        "reconnect": true,
        "reconnect_interval": 1000,
        "rest_options": Object {},
        "rpc_id": 0,
        "webSocketFactory": [Function],
      },
      "_rpcWebSocketConnected": false,
      "_rpcWebSocketGeneration": 0,
      "_rpcWebSocketHeartbeat": null,
      "_rpcWebSocketIdleTimeout": null,
      "_rpcWsEndpoint": "wss://api.devnet.solana.com/",
      "_subscriptionCallbacksByServerSubscriptionId": Object {},
      "_subscriptionDisposeFunctionsByClientSubscriptionId": Object {},
      "_subscriptionsAutoDisposedByRpc": Set {},
      "_subscriptionsByHash": Object {},
    },
    "networkIdentifier": "devnet",
    "publicKeys": Object {},
  },
  "usersProgram": null,
}
`;
